src/iop/filmicrgb.c

```diff
@@ -423,117 +493,905 @@ static inline float linear_saturation(const float x, const float luminance, cons
+  uniform(self, piece, roi_in, roi_out)
+#endif
+void process(dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *piece, const void *const restrict ivoid, void *const restrict ovoid,
+             const dt_iop_roi_t *const roi_in, const dt_iop_roi_t *const roi_out)
+{
+  const dt_iop_filmicrgb_data_t *const data = (dt_iop_filmicrgb_data_t *)piece->data;
+  const dt_iop_order_iccprofile_info_t *const work_profile = dt_ioppr_get_pipe_work_profile_info(piece->pipe);
+
+  if(piece->colors != 4)
+  {
+    dt_control_log(_("filmic works only on RGB input"));
+    return;
+  }
 
-      float DT_ALIGNED_PIXEL ratios[4];
-      float norm = get_pixel_norm(pix_in, variant, work_profile);
+  const size_t ch = 4;
 
-      norm = (norm < 1.52587890625e-05f) ? 1.52587890625e-05f : norm; // norm can't be < to 2^(-16)
+  /** The log2(x) -> -INF when x -> 0
+  * thus very low values (noise) will get even lower, resulting in noise negative amplification,
+  * which leads to pepper noise in shadows. To avoid that, we need to clip values that are noise for sure.
+  * Using 16 bits RAW data, the black value (known by rawspeed for every manufacturer) could be used as a threshold.
+  * However, at this point of the pixelpipe, the RAW levels have already been corrected and everything can happen with black levels
+  * in the exposure module. So we define the threshold as the first non-null 16 bit integer
+  */
+
+  float *restrict in = (float *)ivoid;
+  float *const restrict out = (float *)ovoid;
+  float *const restrict mask =  dt_alloc_sse_ps(roi_out->width * roi_out->height);
 
-      // Save the ratios
-      for(int c = 0; c < 3; c++) ratios[c] = pix_in[c] / norm;
+  // build a mask of clipped pixels
+  const float normalize = data->reconstruct_feather / data->reconstruct_threshold;
+  const int recover_highlights = mask_clipped_pixels(in, mask, normalize, data->reconstruct_feather, roi_out->width, roi_out->height, 4);
 
-      // Sanitize the ratios
-      const float min_ratios = fminf(fminf(ratios[0], ratios[1]), ratios[2]);
-      if(min_ratios < 0.0f) for(int c = 0; c < 3; c++) ratios[c] -= min_ratios;
+  // display mask and exit
+  if(self->dev->gui_attached && piece->pipe->type == DT_DEV_PIXELPIPE_FULL && mask)
+  {
+    dt_iop_filmicrgb_gui_data_t *g = (dt_iop_filmicrgb_gui_data_t *)self->gui_data;
 
-      // Log tone-mapping
-      norm = log_tonemapping(norm, data->grey_source, data->black_source, data->dynamic_range);
+    if(g->show_mask)
+    {
+      display_mask(mask, out, roi_out->width, roi_out->height, ch);
+      dt_free_align(mask);
+      return;
+    }
+  }
 
-      // Get the desaturation value based on the log value
-      const float desaturation = filmic_desaturate(norm, data->sigma_toe, data->sigma_shoulder, data->saturation);
+  float *const restrict reconstructed = dt_alloc_sse_ps(roi_out->width * roi_out->height * ch);
 
-      for(int c = 0; c < 3; c++) ratios[c] *= norm;
+  if(recover_highlights && mask && reconstructed)
```

Comment:
could the diff I'm seeing on the integration test be there ? shouldn't this be activated only if using new v4 algo ?

In legacy_param() we have:

`    n->reconstruct_threshold = 3.0f; // for old edits, this ensures clipping threshold >> white level, so it's a no-op`

Is the no-op really the case ?

Reply:
There is no option to disable recovery, only you can use crazy high threshold for the highlight masking, so it is practically disabled. That could indeed be the cause, in that case this value needs to be increased (although it's actually a good thing, anyway, that highlights are a bit blurred where they clip).